src/widgets/keyboard/lv_keyboard.c

```diff
@@ -371,14 +582,314 @@ void lv_keyboard_def_event_cb(lv_event_t * e)
+            new_pinyin_str[i] = pinyin_str[i];
+        }
+        else {
+            return NULL;
+        }
+    }
+
+    if(keyboard->dict == NULL) {
+        LV_LOG_ERROR("keyboard pinyin dict is NULL");
+        return NULL;
+    }
+
+    lv_keyboard_pinyin_dict_t * pinyin_head = (lv_keyboard_pinyin_dict_t *)keyboard->dict[*new_pinyin_str - 'a'];
+    lv_keyboard_pinyin_dict_t * pinyin_tail = (lv_keyboard_pinyin_dict_t *)keyboard->dict[*new_pinyin_str - 'a' + 1];
+
+    size_t i = 0;
+    for(; pinyin_head < pinyin_tail; pinyin_head++) {
+        for(i = 0; i < pinyin_str_len - 1; i++) {
+            if((*pinyin_head).pinyin_list == NULL) return NULL;
+
+            if(*((*pinyin_head).pinyin_list + i) != new_pinyin_str[i + 1]) break;
+        }
+
+        /*perfect match*/
+        if(i == (pinyin_str_len - 1)) {
+            /*The Chinese character in UTF-8 encoding format is 3 bytes*/
+            *candidate_num = strlen((const char *)(*pinyin_head).pinyin_mb_list) / 3;
+
+            return (char *)(*pinyin_head).pinyin_mb_list;
+        }
+    }
+
+    return NULL;
+}
+
+static void lv_keyboard_proc_pinyin(lv_keyboard_t * keyboard)
+{
+    char * input_pinyin = lv_label_get_text(keyboard->pinyin_label);
+    if(lv_strlen(input_pinyin) == 0) {
+        return;
+    }
+
+    size_t candidate_num = 0;
+    char * candidate_str = lv_keyboard_search_chinese(&candidate_num, keyboard, input_pinyin);
+
+    lv_keyboard_free_candidate_btnm_map(keyboard);
+    keyboard->candidate_num = candidate_num;
+
+    keyboard->candidate_btnm_map = (char **)lv_malloc((candidate_num + 1) * sizeof(char *));
+    if(keyboard->candidate_btnm_map == NULL) {
+        return;
+    }
+
+    for(size_t i = 0; i < candidate_num; i++) {
+        keyboard->candidate_btnm_map[i] = (char *)lv_malloc(4 * sizeof(char));
+        if(keyboard->candidate_btnm_map[i] == NULL) {
+            lv_keyboard_free_candidate_btnm_map(keyboard);
+            return;
+        }
+
+        lv_memcpy(keyboard->candidate_btnm_map[i], candidate_str + i * 3, 3);
+        keyboard->candidate_btnm_map[i][3] = '\0';
+    }
+
+    keyboard->candidate_btnm_map[keyboard->candidate_num] = "";
+
+    lv_buttonmatrix_set_map(keyboard->candidate_btnm, (const char **)keyboard->candidate_btnm_map);
+
+    int32_t height = lv_obj_get_height(keyboard->candidate_btnm);
+    lv_obj_set_width(keyboard->candidate_btnm, height * keyboard->candidate_num);
+}
+
+#if LV_KEYBOARD_PINYIN_USE_EXT_BUTTONMATRIX
+static void lv_keyboard_proc_ext_candidate_btnm_map(lv_keyboard_t * keyboard)
+{
+    int32_t cont_width = lv_obj_get_width(keyboard->candidate_btnm_cont);
+    int32_t btnm_width = lv_obj_get_width(keyboard->candidate_btnm);
+    if(btnm_width <= cont_width) {
+        lv_obj_add_flag(keyboard->ext_btn, LV_OBJ_FLAG_HIDDEN);
+        return;
+    }
+
+    lv_obj_remove_flag(keyboard->ext_btn, LV_OBJ_FLAG_HIDDEN);
+
+    int32_t width = (btnm_width / keyboard->candidate_num);
```

Comment:
Please confirm if it's needed to check divide zero here. And similar for below division.

Reply:
> Please confirm if it's needed to check divide zero here. And similar for below division.

No need to check here, But add the following check
```c
size_t candidate_num = 0;
char * candidate_str = lv_keyboard_search_chinese(&candidate_num, keyboard, input_pinyin);
if(candidate_str == NULL || lv_strlen(candidate_str) == 0 || candidate_num == 0) return; // add
```